Add benchmarks, faster batch_insert.

Pratyush opened the pull request:
Hi,

This is a PR for some perf improvements: 

1. A new `FromIterator` impl that collects the iterator into a `Vec`, sorts it by the key, dedups by the key, and constructs a new `FlatMap` from this. This is *much* faster for large inputs than inserting each pair one by one because the latter requires moving elements down the vector whenever an insertion occurs.

2. I added `#[inline]` attributes to small functions.

I also added some benchmarks comparing the map to BTreeMap from the stdlib, and added myself to the contributors list.

toffaletti commented on benches/flat_map.rs:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    // key_vals.sort_by(|pair1, pair2|  pair1.0.cmp(&pair2.0));
+    // key_vals.dedup();
```
I'm not a fan of committing commented out code. Can you please remove this and the two `// let key = rand::random::<Key>();` above?

toffaletti commented on src/flat_map.rs:
```diff
@@ -65,10 +60,12 @@ pub struct Values<'a, K: 'a, V: 'a> {
 }
 
 impl<K, V> FlatMap<K, V> {
+    #[inline]
```
I went looking at the std lib for its usage of #[inline] on collections. For example, https://github.com/rust-lang/rust/blob/master/src/liballoc/btree/map.rs only has 5 usage of inline attribute. On the other hand, https://github.com/rust-lang/rust/blob/master/src/liballoc/vec.rs which is closer to what we're doing here has 50.

I went looking for the cost of using inline and it looks like it can slow compile times and increase binary sizes (maybe that doesn't matter here because this is generic code). Another option instead of aggressively using inline in a library is letting the application author use LTO: https://internals.rust-lang.org/t/when-should-i-use-inline/598

Just want to make sure you're aware of the costs before we decide this is the right thing.

Pratyush commented:
I pushed some updates:

1. Fixed the benchmark code and added a new `from_iter` benchmark.
2. Removed `#[inline]` attributes because they weren't giving any speed-ups on the benchmarks. Perhaps the compiler is smart enough to know when to inline function calls in our case.
